### ISSUE TITLE
Please consider a native ARM64 version for Windows

### ISSUE THREAD

**Issue by MagnusJohansson** [marticliment/UniGetUI#2407]
### Please confirm these before moving forward.

- [X] I have searched for my feature proposal and have not found a work-in-progress/duplicate/resolved/discarded issue.
- [X] This proposal is a completely new feature. If you want to suggest an improvement or an enhancement, please use [this template](https://github.com/marticliment/WingetUI/issues/new?assignees=marticliment&labels=enhancement&projects=&template=enhancement-improvement.yml&title=%5BENHANCEMENT%5D+%28Enter+your+description+here%29).

### Describe the new feature

I would love to have a native ARM64 build, instead of running on the x64 emulator.
This would improve performance, and more importantly save battery.
Screenshot from UniGetUI v3.1.0:
![image](https://github.com/marticliment/UniGetUI/assets/674909/c88dc3ee-8739-4455-b872-b80f6cdf5b0e)

### Describe how this new feature could help users

It would improve the end user experience:
- Better performance du to not have to be run via the emulation layer
- Improved energy consumption
- Improved responsiveness
- Align with trends; I would assume that many WinGetUI/UniGetUI users are technical and more likely to be running a ARM64 based device than any other audience

**Comment by marticliment** [comment]
I don't have an arm64 machine (yet), and I don't want to build arm64 with my x64 computer and release it without testing.

However, I will see what I can do

**Comment by Vascro** [comment]
> I don't have an arm64 machine (yet), and I don't want to build arm64 with my x64 computer and release it without testing.
> 
> However, I will see what I can do

If you'd like someone to test, I own an Arm64 snapdragon x elite laptop and I'd be happy to assist. I also own a SQ3 surface pro 9 so we'd have coverage with multiple arm based cpu's.

**Comment by christian-bendiksen** [comment]
Would like to see this as well. Running Surface Laptop 7 x elite. Could probably test a build as well.

**Comment by JaronrH** [comment]
I have a Surface Pro X and can test it out for you too!

**Comment by CanePlayz** [comment]
I can test it too. :P Jokes aside, I'd also love to see this.

**Comment by AndrewCLion** [comment]
I have a 32GB Snapdragon X Elite laptop and Visual Studio Professional, I had a quick look at compiling but there were a number of libraries "missing". So, a tip of the hat to hmartinez82 who must have put more effort in than I did. I agree, it's not just a case of getting it to compile and run, there's all the "which architecture" questions to be answered or decided.

**Comment by Vascro** [comment]
> I can get it to build, the trick part will be the following when running on ARM64: Given that UniGetUI can detect the architecture of the installing application: If the installed app is x64, should we upgrade to a newer x64 or try to install the ARM64 version if available?

Would it be possible to make it either a checkbox before installing updates, or have unigetui prompt yes/no to replace with arm64 native build if it detects one?

**Comment by marticliment** [comment]
I appreciate your efforts, but there won't be an arm64 version in the near future.

The problem here is that while UniGetUI may build under arm64 systems, (in fact, it has no theoretical reason for it to not build) I cannot publish a version I have not tested, and I can't either rely on a third-party to do the builds for me. It would be too unreliable, and I would be taking a very high risk.

Thanks for the testing, and sorry.

**Comment by flanter21** [comment]
> I appreciate your efforts, but there won't be an arm64 version in the near future.
> 
> The problem here is that while UniGetUI may build under arm64 systems, (in fact, it has no theoretical reason for it to not build) I cannot publish a version I have not tested, and I can't either rely on a third-party to do the builds for me. It would be too unreliable, and I would be taking a very high risk.
> 
> Thanks for the testing, and sorry.

Could you use GitHub actions or something? UniGetUI slows my entire system down.

**Comment by marticliment** [comment]
> Could you use GitHub actions or something?

The issue is the same. I could build UniGetUI, but I wouldn't be able to test it

**Comment by flanter21** [comment]
> > Could you use GitHub actions or something?
> 
> The issue is the same. I could build UniGetUI, but I wouldn't be able to test it

Ah I see now sorry. I guess you could use qemu to run arm64 windows (w11 iot enterprise ltsc 2024 would be lightweight) but I can see that would probably be a lot of effort for not a lot of users so fair enough.

**Comment by hmartinez82** [comment]
Ok. So instructions to building for ARM64. They can be executed from both an ARM64 Windows machine or an x64 one (where you can then copy the files to an ARM64 machine).

Pre-Requisites:
- Windows 10 SDK, version 2004 (10.0.19041.0) - All the C++ and UWP checkboxes
- Windows SDK for Windows 11 (10.0.26100.1742)  - All the C++ and UWP checkboxes
- .NET 8 SDK

Probably because there are two projects using CsWinRT  that target different versions of the SDK. They can all be acquired from https://developer.microsoft.com/en-us/windows/downloads/sdk-archive/

Steps:
- Open the `src\UniGetUI.sln` solution in Visual Studio 2022
- Right click the solution -> Configuration Manager
- Click the `Active solution platform:` combobox and select `<New...>`
- In the top combobox, select `ARM64`, the in bottom combobox select `x64` and make sure that the `Create new project platforms` checkbox is selected
- Save the solution and close Visual Studio
- In a command prompt in the top of the repository type: `dotnet publish src\UniGetUi\UniGetUi.csproj --configuration Release --runtime win-arm64 --self-contained`
- Binaries will be in the `src\UniGetUI\bin\Release\net8.0-windows10.0.26100.0\win-arm64` folder

I was wondering if the existing code would force the winget packages to be built for x64, but they winget is being called without forcing a specific architecture, thus is defaults to ARM64.

**Comment by hmartinez82** [comment]
@marticliment Given the stops above are super simple and cross compilation is very trivial, could you perhaps make an Experimental build?

Based on that batch file, the build line would be probably be:
`dotnet publish src/UniGetUI/UniGetUI.csproj /noLogo /property:Configuration=Release /property:Platform=arm64 -v m`

**Comment by marticliment** [comment]
Yes, but the issue is not building UniGetUI, but rather being able to test it.

**Comment by flanter21** [comment]
For anyone who wants to build, you can use [this Github Action](https://github.com/flanter21/UniGetUI-arm64/blob/main/.github/workflows/build-no-tests.yaml).

Note that it's not fully there, as [InstallerExtras/CodeDependencies.iss](https://github.com/marticliment/UniGetUI/blob/main/InstallerExtras/CodeDependencies.iss) seems to only support x64 and I've [modified](https://github.com/flanter21/UniGetUI-arm64/blob/main/CodeDependencies.iss) it to support both x86 and, where possible, ARM.

The repo also has x64 dependencies in the repo, such as

- [src/UniGetUI.PackageEngine.Managers.Chocolatey/choco-cli](https://github.com/marticliment/UniGetUI/tree/main/src/UniGetUI.PackageEngine.Managers.Chocolatey/choco-cli)
- [src/UniGetUI/Assets/Utilities](https://github.com/marticliment/UniGetUI/tree/main/src/UniGetUI/Assets/Utilities)
- [src/UniGetUI.PackageEngine.Managers.WinGet/winget-cli_x64](https://github.com/marticliment/UniGetUI/tree/main/src/UniGetUI.PackageEngine.Managers.WinGet/winget-cli_x64)

These should ideally change according to the architecture of the build, but I'll leave it there for now.

**Comment by hmartinez82** [comment]
@flanter21 Did you get `gsudo` working? I think that the gsudo code checks that the parent process' image is signed with a specific signature. In other words, using batch elevation doesn't seem to work.

**Comment by marticliment** [comment]
> [@flanter21](https://github.com/flanter21?rgh-link-date=2025-03-30T04%3A10%3A22.000Z) Did you get `gsudo` working? I think that the gsudo code checks that the parent process' image is signed with a specific signature. In other words, using batch elevation doesn't seem to work.

No, it will not work.
However, you can enable `gsudo` instead of `UniGetUI Elevator` from the experimental settings. GSudo does not check for digital signatures, so batching should work then

**Comment by hmartinez82** [comment]
@marticliment That did it, thank you :)

Btw, Microsoft **finally** released Windows ARM64 runners for all Github projects!

**Comment by vectorrilke** [comment]
> For anyone who wants to build, you can use [this Github Action](https://github.com/flanter21/UniGetUI-arm64/blob/main/.github/workflows/build-no-tests.yaml).
> 
> Note that it's not fully there, as [InstallerExtras/CodeDependencies.iss](https://github.com/marticliment/UniGetUI/blob/main/InstallerExtras/CodeDependencies.iss) seems to only support x64 and I've [modified](https://github.com/flanter21/UniGetUI-arm64/blob/main/CodeDependencies.iss) it to support both x86 and, where possible, ARM.
> 
> The repo also has x64 dependencies in the repo, such as
> 
>     * [src/UniGetUI.PackageEngine.Managers.Chocolatey/choco-cli](https://github.com/marticliment/UniGetUI/tree/main/src/UniGetUI.PackageEngine.Managers.Chocolatey/choco-cli)
> 
>     * [src/UniGetUI/Assets/Utilities](https://github.com/marticliment/UniGetUI/tree/main/src/UniGetUI/Assets/Utilities)
> 
>     * [src/UniGetUI.PackageEngine.Managers.WinGet/winget-cli_x64](https://github.com/marticliment/UniGetUI/tree/main/src/UniGetUI.PackageEngine.Managers.WinGet/winget-cli_x64)
> 
> 
> These should ideally change according to the architecture of the build, but I'll leave it there for now. It should probably also be extended to cover https://github.com/marticliment/GSudo-for-UniGetUI
> 
> It's insane how much of a difference it makes though. It's super smooth now, when before it would cause lag just on opening or scrolling (and my laptop wasn't cheap either)

Could you publish it as a third-party experimental build?

**Comment by OliverLx** [comment]
I can give you RDP user access to an ARM64 Azure VM that I just created. Contact me directly if you like.
Yesterday, I ordered a https://www.asus.com/de/laptops/for-home/vivobook/asus-vivobook-14-x1407q for 499 € my wife and wanted to test upfront, saw a Lonove at a customer but would never ever enter my passwords on systems with firmware and updates from "some" countries...

**Comment by marticliment** [comment]
Hello @OliverLx, Thanks for the offer, but I am afraid I can't take it.

**Comment by OliverLx** [comment]
Hello @marticliment, you're welcome! I understand that everything takes time and effort. I'm sure that one day you'll own such a device too, and you can take a look then. :-) Thank you for your fantastic product! I have just installed and tested your x64 version via Winget and/or the Microsoft Store, and it works perfectly as usual.

**Comment by talynone** [comment]
Looks like a Windows ARM version is now out!

https://github.com/Devolutions/UniGetUI/releases

This was an error during install when it tried to install the x64 redistributable package, I just clicked ignore and let it finish, it runs.

After first restart of app it ask to install cargo dependencies, I let it do it, seems to finish ok. But any time I try to actually update an app I get the following error, no matter what method I try, admin, interactive, etc...:

<img width="631" height="509" alt="Image" src="https://github.com/user-attachments/assets/29ff5609-96af-4cfd-bbee-d94cbf76460c" />

**Comment by vectorrilke** [comment]
I am getting same errors:

```
Package update operation for Package=Adobe.Acrobat.Reader.64-bit with Manager=Winget
Installation options: <InstallOptions instance (only non-default values are shown)
	OverridesNextLevelOpts: True>
Overriden options: <Scope=;RunAsAdministrator=True;WG_SpecifyVersion=;PS_NoScope=False>
Version: 25.001.21184 -> 25.001.21288
Starting operation...
Executing process with StartInfo:
 - FileName: "C:\Program Files\UniGetUI\Assets\Utilities\UniGetUI Elevator.exe"
 - Arguments: ""C:\Users\vector\AppData\Local\Microsoft\WindowsApps\winget.exe"  update --id "Adobe.Acrobat.Reader.64-bit" --exact --source winget --accept-source-agreements --disable-interactivity --silent --include-unknown --accept-package-agreements --force --architecture x64"
Start Time: "15.03.2026 22:15:41"
Warning: W_HELPER_DLL_NOT_FOUND
Error: Caller process failed integrity verification. Exiting.
End Time: "15.03.2026 22:15:41"
Process return value: "999" (0x3E7)
Adobe Acrobat Reader (64-bit) could not be updated
```

**Comment by vectorrilke** [comment]
Also there is an error during installation of arm64 version of UnigetUI - installer is trying to download VC redist x86 version...

**Comment by codesimplicity** [comment]
Same at a friend's new machine, would love to hear about (temporary) solutions.
Would installing the x64 version be the best bet for now?

My main problem is that I won't have direct access regularly, so fiddling with workarounds would be annoying - have fallen back to manual updates for now and 🙏

**Comment by hmartinez82** [comment]
The issue is that `getfilesiginforedist.dll` is missing from the ARM64 installation, but is present in the x64 installation.

I downloaded the ARM64 DLL from within the https://www.nuget.org/packages/Microsoft.Security.Extensions#dependencies-body-tab NuGet package and extracted it to `<installed folder>\Assets\Utilities`.

It then finds the file, but now UniGetUI fails with `W_HELPER_DLL_HASH_MISMATCH` which means this need to be fixed in the source code. I'll try to cook up a PR sometime later this week.

**Comment by hmartinez82** [comment]
Hopefully now addressed https://github.com/Devolutions/UniGetUI/pull/4567 :)

**Comment by hmartinez82** [comment]
I can confirm that the MSVC installation during the installer doesn't fail on ARM64 and the Elevator is also working:

<img width="1859" height="336" alt="Image" src="https://github.com/user-attachments/assets/1b8e5264-9ffd-4ab9-8e9c-17f8483a7bfb" />

**Comment by hmartinez82** [comment]
For anyone interested in installing the fixed ARM64 build, there's a signed build in the Artifact section (scroll all the way down) at https://github.com/Devolutions/UniGetUI/actions/runs/24205910846

**Comment by vectorrilke** [comment]
> For anyone interested in installing the fixed ARM64 build, there's a signed build in the Artifact section (scroll all the way down) at https://github.com/Devolutions/UniGetUI/actions/runs/24205910846

Installer is still downloading and installing x64 VC redist. Is it by design?

**Comment by hmartinez82** [comment]
> > For anyone interested in installing the fixed ARM64 build, there's a signed build in the Artifact section (scroll all the way down) at https://github.com/Devolutions/UniGetUI/actions/runs/24205910846
> 
> Installer is still downloading and installing x64 VC redist. Is it by design?

Yes, it's by design. There was a change indeed, just for a modern x64 installer. The title is x64, but it actually installs the ARM64 on ARM64 Windows. See the x64 comment at: https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170#latest-supported-redistributable-version

**Comment by talynone** [comment]
Seems to be working so far, a little tip:

You can modify your WinGet settings to prefer ARM64 globally.

Open your settings file with the command: `winget settings`

Add or update the `installBehavior `section to prioritize ARM64:

```
"installBehavior": {
    "preferences": {
        "architectures": ["arm64", "x64"]
    }
}
```
This tells WinGet: "Always look for a native ARM64 installer first. If you can't find one, only then try the x64 version." This should stop it from defaulting to the x86/x64 version during mass updates for some applications.

**Comment by vectorrilke** [comment]
@talynone when I add this to the settings, will it solve the issue with preferred architecture in unigetui? Current behaviour is if I set arch to arm64 and package is x64, it fails to update unless I specifically disable Default settings per package in installation options.

**Comment by talynone** [comment]
> [talynone (@talynone)](https://github.com/talynone) when I add this to the settings, will it solve the issue with preferred architecture in unigetui? Current behaviour is if I set arch to arm64 and package is x64, it fails to update unless I specifically disable Default settings per package in installation options.

Not sure, the problem I was having was UniGetUi was overriding my Notepad++ ARM64 with X64 version even though the download url in the info was set to the ARM64 version. I used the above winget config then forced an install and it seems to install the ARM64 version. Will need more testing.

**Comment by hmartinez82** [comment]
There are also the corner cases, like LLVM/Clang . Sometimes the Windows ARM64 installer takes a day or two to be built compared to everything else. Here's an example:
<img width="1864" height="1850" alt="Image" src="https://github.com/user-attachments/assets/1a166953-ea3d-4867-8b4f-fec57b700f77" />

I've seen these difference in  days sometime be up to 5 days!
